File: frontend/HobbyStreamerCMS/src/types/asset.ts
export interface Asset {
  id: string;
  slug: string;
  title?: string;
  description?: string;
  type: AssetType;
  genre?: string;
  genres?: string[];
  tags?: string[];
  status?: string;
  createdAt: string;
  updatedAt: string;
  metadata?: Record<string, any>;
  ownerId?: string;
  parent?: Asset;
  children?: Asset[];
  buckets?: Bucket[];
  videos?: Video[];
  images?: Image[];
  publishRule?: PublishRule;
}

export interface PublishRule {
  publishAt?: string;
  unpublishAt?: string;
  regions?: string[];
  ageRating?: string;
}

export enum VideoType {
  MAIN = 'main',
  TRAILER = 'trailer',
  BEHIND_THE_SCENES = 'behind',
  INTERVIEW = 'interview'
}

export enum VideoFormat {
  RAW = 'raw',
  HLS = 'hls',
  DASH = 'dash'
}

export interface S3Object {
  bucket: string;
  key: string;
  url: string;
}

export interface StreamInfo {
  downloadUrl?: string;
  cdnPrefix?: string;
  url?: string;
}

export interface Video {
  id: string;
  type: VideoType;
  format: string;
  storageLocation: S3Object;
  width?: number;
  height?: number;
  duration?: number;
  bitrate?: number;
  codec?: string;
  size?: number;
  contentType?: string;
  streamInfo?: StreamInfo;
  metadata?: Record<string, any>;
  status?: string;
  thumbnail?: Image;
  createdAt: string;
  updatedAt: string;
}

export enum ImageType {
  THUMBNAIL = 'thumbnail',
  POSTER = 'poster',
  BANNER = 'banner',
  HERO = 'hero',
  LOGO = 'logo',
  SCREENSHOT = 'screenshot',
  BEHIND_THE_SCENES = 'behind_scenes',
  INTERVIEW = 'interview'
}

export interface Image {
  id: string;
  fileName: string;
  url: string;
  type: ImageType;
  storageLocation?: S3Object;
  width?: number;
  height?: number;
  size?: number;
  contentType?: string;
  metadata?: Record<string, any>;
  createdAt: string;
  updatedAt: string;
}


export enum AssetStatus {
  DRAFT = 'draft',
  PUBLISHED = 'published'
}


export enum VideoStatus {
  PENDING = 'pending',
  ANALYZING = 'analyzing',
  TRANSCODING = 'transcoding',
  READY = 'ready',
  FAILED = 'failed'
}


export enum AssetGenre {
  ACTION = 'action',
  DRAMA = 'drama',
  COMEDY = 'comedy',
  HORROR = 'horror',
  SCI_FI = 'sci_fi',
  ROMANCE = 'romance',
  THRILLER = 'thriller',
  FANTASY = 'fantasy',
  DOCUMENTARY = 'documentary',
  MUSIC = 'music',
  NEWS = 'news',
  SPORTS = 'sports',
  KIDS = 'kids',
  EDUCATIONAL = 'educational'
}


export const formatGenreName = (genre: string): string => {
  return genre.replace('_', ' ').toUpperCase();
};



export enum AssetType {
  MOVIE = 'movie',
  TV_SHOW = 'tv_show',
  SERIES = 'series',
  SEASON = 'season',
  EPISODE = 'episode',
  DOCUMENTARY = 'documentary',
  SHORT = 'short',
  TRAILER = 'trailer',
  BONUS = 'bonus',
  BEHIND_SCENES = 'behind_scenes',
  INTERVIEW = 'interview',
  MUSIC_VIDEO = 'music_video',
  PODCAST = 'podcast',
  LIVE = 'live'
}

export interface Bucket {
  id: string;
  key: string;
  name: string;
  description?: string;
  type: BucketType;
  status?: string;
  createdAt: string;
  updatedAt: string;
  assets?: Asset[];
  ownerId?: string;
}

export enum BucketType {
  COLLECTION = 'collection',
  PLAYLIST = 'playlist',
  CATEGORY = 'category'
}

export enum BucketStatus {
  ACTIVE = 'active',
  INACTIVE = 'inactive',
  DRAFT = 'draft'
}

export interface AssetPage {
  items: Asset[];
  nextKey?: string;
}

export interface BucketPage {
  items: Bucket[];
  nextKey?: string;
}

export interface AssetInput {
  slug: string;
  title?: string;
  description?: string;
  type: AssetType;
  genre?: string;
  genres?: string[];
  tags?: string[];
  metadata?: string;
  ownerId?: string;
  parentId?: string;
}

export interface BucketInput {
  key: string;
  name: string;
  description?: string;
  type: BucketType;
  status?: string;
}

export interface AssetCreateDTO {
  title?: string;
  slug?: string;
  description?: string;
  type: AssetType;
  genre?: string;
  genres?: string[];
  tags?: string[];
  metadata?: Record<string, any>;
  ownerId?: string;
  parentId?: string;
}

export interface AssetUpdateDTO {
  title?: string;
  description?: string;
  type?: AssetType;
  genre?: string;
  genres?: string[];
  tags?: string[];
  metadata?: Record<string, any>;
  ownerId?: string;
  parentId?: string;
  status?: string;
}

export interface UploadResponse {
  uploadUrl: string;
  assetId: number;
  fileName: string;
}

export interface AssetUpdateInput {
  title?: string | null;
  description?: string | null;
  type?: AssetType | null;
  genre?: string | null;
  genres?: string[] | null;
  tags?: string[] | null;
  metadata?: string | null;
  ownerId?: string | null;
  parentId?: string | null;
  clearFields?: string[];
} 

File: backend/streaming-api/internal/model/asset.go
package model

import "time"

type Asset struct {
	ID          string       `json:"id"`
	Slug        string       `json:"slug"`
	Title       *string      `json:"title,omitempty"`
	Description *string      `json:"description,omitempty"`
	Type        string       `json:"type"`
	Genre       *string      `json:"genre,omitempty"`
	Genres      []string     `json:"genres,omitempty"`
	Tags        []string     `json:"tags,omitempty"`
	Status      *string      `json:"status,omitempty"`
	CreatedAt   time.Time    `json:"createdAt"`
	UpdatedAt   time.Time    `json:"updatedAt"`
	Metadata    *string      `json:"metadata,omitempty"`
	OwnerID     *string      `json:"ownerId,omitempty"`
	Videos      []Video      `json:"videos,omitempty"`
	Images      []Image      `json:"images,omitempty"`
	PublishRule *PublishRule `json:"publishRule,omitempty"`
}

type Video struct {
	ID              string      `json:"id"`
	Type            string      `json:"type"`
	Format          string      `json:"format"`
	StorageLocation S3Object    `json:"storageLocation"`
	Width           *int        `json:"width,omitempty"`
	Height          *int        `json:"height,omitempty"`
	Duration        *float64    `json:"duration,omitempty"`
	Bitrate         *int        `json:"bitrate,omitempty"`
	Codec           *string     `json:"codec,omitempty"`
	Size            *int        `json:"size,omitempty"`
	ContentType     *string     `json:"contentType,omitempty"`
	StreamInfo      *StreamInfo `json:"streamInfo,omitempty"`
	Metadata        *string     `json:"metadata,omitempty"`
	Status          *string     `json:"status,omitempty"`
	Thumbnail       *Image      `json:"thumbnail,omitempty"`
	CreatedAt       time.Time   `json:"createdAt"`
	UpdatedAt       time.Time   `json:"updatedAt"`
}

type Image struct {
	ID              string      `json:"id"`
	FileName        string      `json:"fileName"`
	URL             string      `json:"url"`
	Type            string      `json:"type"`
	StorageLocation *S3Object   `json:"storageLocation,omitempty"`
	Width           *int        `json:"width,omitempty"`
	Height          *int        `json:"height,omitempty"`
	Size            *int        `json:"size,omitempty"`
	ContentType     *string     `json:"contentType,omitempty"`
	StreamInfo      *StreamInfo `json:"streamInfo,omitempty"`
	Metadata        *string     `json:"metadata,omitempty"`
	CreatedAt       time.Time   `json:"createdAt"`
	UpdatedAt       time.Time   `json:"updatedAt"`
}

type S3Object struct {
	Bucket string `json:"bucket"`
	Key    string `json:"key"`
	URL    string `json:"url"`
}

type StreamInfo struct {
	DownloadURL *string `json:"downloadUrl,omitempty"`
	CDNPrefix   *string `json:"cdnPrefix,omitempty"`
	URL         *string `json:"url,omitempty"`
}

type PublishRule struct {
	IsPublic    bool       `json:"isPublic"`
	PublishAt   *time.Time `json:"publishAt,omitempty"`
	UnpublishAt *time.Time `json:"unpublishAt,omitempty"`
	Regions     []string   `json:"regions,omitempty"`
	AgeRating   *string    `json:"ageRating,omitempty"`
}

type AssetPage struct {
	Items   []Asset `json:"items"`
	NextKey *string `json:"nextKey,omitempty"`
}
